### PR TITLE
imp: remove `height` from `header` and merge `header` with `misbehaviour`

### PR DESCRIPTION
**Pull request by crodriguezvega**
## Description

Work in progress. Still needs new versions of contracts for tests to pass.

closes: #3958 

### Commit Message / Changelog Entry

```bash
type: commit message
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [x] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/light-clients/08-wasm/types/vm.go [hunk 1]
```diff
@@ -40,6 +40,16 @@ func (r contractResult) Error() string {
...
 
+type checkForMisbehaviourExecuteResult struct {
+	contractResult
+	FoundMisbehaviour bool `json:"found_misbehaviour"`
+}
+
+type updateStateExecuteResult struct {
+	contractResult
+	Heights []exported.Height `json:"heights"`
+}
```
Comment (crodriguezvega) [inline modules/light-clients/08-wasm/types/vm.go:51]: Added these types for the different results.

File: modules/light-clients/08-wasm/types/update.go [hunk 2]
```diff
@@ -13,11 +14,7 @@ var _ exported.ClientState = (*ClientState)(nil)
 
 type (
 	verifyClientMessageInnerPayload struct {
-		ClientMessage clientMessage `json:"client_message"`
-	}
-	clientMessage struct {
-		Header       *Header       `json:"header,omitempty"`
-		Misbehaviour *Misbehaviour `json:"misbehaviour,omitempty"`
+		ClientMessage *ClientMessage `json:"client_message"`
```
Comment (crodriguezvega) [inline modules/light-clients/08-wasm/types/update.go:17]: This is how the payload to the contract call looks after merging `Header` and `Misbehaviour`.

File: modules/light-clients/08-wasm/types/client_message.go [hunk 3]
```diff
@@ -6,15 +6,15 @@ import (
 	"github.com/cosmos/ibc-go/v7/modules/core/exported"
 )
 
-var _ exported.ClientMessage = &Header{}
+var _ exported.ClientMessage = &ClientMessage{}
 
-// ClientType defines that the Header is a Wasm client consensus algorithm
-func (h Header) ClientType() string {
+// ClientType defines that the client message is a Wasm client consensus algorithm
+func (h ClientMessage) ClientType() string {
```
Comment (damiannolan) [inline modules/light-clients/08-wasm/types/client_message.go:12]: we can change `h` -> `c`/`cm`?

```suggestion
func (c ClientMessage) ClientType() string {
```

File: modules/light-clients/08-wasm/types/client_message.go [hunk 4]
```diff
@@ -6,15 +6,15 @@ import (
...
 	return exported.Wasm
 }
 
-// ValidateBasic defines a basic validation for the wasm client header.
-func (h Header) ValidateBasic() error {
+// ValidateBasic defines a basic validation for the wasm client message.
+func (h ClientMessage) ValidateBasic() error {
```
Comment (damiannolan) [inline modules/light-clients/08-wasm/types/client_message.go:17]: ditto 

```suggestion
func (c ClientMessage) ValidateBasic() error {
```

File: modules/light-clients/08-wasm/types/client_message_test.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+			&types.ClientMessage{
+				Data: []byte{},
+			},
+			false,
+		},
+	}
+
+	for _, tc := range testCases {
+		suite.Run(tc.name, func() {
+			clientMessage := tc.clientMessage
```
Comment (damiannolan) [inline modules/light-clients/08-wasm/types/client_message_test.go:39]: nit: redundant var, but feel free to ignore, its probably fine

File: modules/light-clients/08-wasm/types/update.go [hunk 6]
```diff
@@ -29,29 +26,23 @@ type (
...
-	case *Header:
-		clientMsgConcrete.Header = clientMsg
-	case *Misbehaviour:
-		clientMsgConcrete.Misbehaviour = clientMsg
-	}
-	inner := verifyClientMessageInnerPayload{
-		ClientMessage: clientMsgConcrete,
+	clientMessage, ok := clientMsg.(*ClientMessage)
+	if !ok {
+		return sdkerrors.Wrapf(ErrInvalid, "expected type: %T, got: %T", &ClientMessage{}, clientMsg)
```
Comment (damiannolan) [inline modules/light-clients/08-wasm/types/update.go:31]: there should be some kind of `ErrInvalidType` perhaps in `ibcerrors`?

File: modules/light-clients/08-wasm/types/update.go [hunk 7]
```diff
@@ -60,30 +51,28 @@ type (
 
 // Client state and new consensus states are updated in the store by the contract
 func (cs ClientState) UpdateState(ctx sdk.Context, cdc codec.BinaryCodec, clientStore sdk.KVStore, clientMsg exported.ClientMessage) []exported.Height {
-	header, ok := clientMsg.(*Header)
+	clientMessage, ok := clientMsg.(*ClientMessage)
 	if !ok {
-		panic(fmt.Errorf("expected type %T, got %T", &Header{}, clientMsg))
+		panic(fmt.Errorf("expected type %T, got %T", &ClientMessage{}, clientMsg))
```
Comment (damiannolan) [inline modules/light-clients/08-wasm/types/update.go:56]: if we're going to panic here, then we should also panic in `CheckForMisbehaviour` and `UpdateStateOnMisbehaviour`  if `clientMsg` is not `*ClientMessage`, right?

File: modules/light-clients/08-wasm/types/vm.go (hunk 1 above)
Comment (damiannolan) [reply modules/light-clients/08-wasm/types/vm.go:51]: same comment from other PR about response vs result naming :)

**Review by damiannolan — COMMENTED** [review]
left some comments / minor suggestions! :) nice work on all these!

File: modules/light-clients/08-wasm/types/update.go (hunk 7 above)
Comment (DimitrisJim) [reply modules/light-clients/08-wasm/types/update.go:56]: I think this mostly followed what is done for tendermint. For CheckForMisbehaviour, returning `false` would call `UpdateState` and result in that panic being hit. I guess panicking could be done here but it wouldn't save us much right?

Comment (damiannolan) [reply modules/light-clients/08-wasm/types/update.go:56]: Agree!

**Pull request merged**